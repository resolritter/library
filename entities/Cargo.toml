[package]
name = "entities"
version = "0.1.0"
authors = ["resolritter <resolritter@users.noreply.github.com>"]
edition = "2018"

[dependencies]
surf = "2.0"
serde = { version = "1", features = ["derive"] }
serde_json = "1"
