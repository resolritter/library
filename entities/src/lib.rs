mod macros;
#[macro_use]
pub mod book;
mod data;
pub mod user;

pub use crate::data::*;
