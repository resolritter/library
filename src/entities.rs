use bastion::prelude::BastionContext;
use serde::{Deserialize, Serialize};
use sqlx::PgPool;
use std::sync::Arc;

pub struct App {
    pub db_pool: &'static Arc<&'static PgPool>,
    pub bastion: &'static Arc<&'static BastionContext>,
}
impl std::fmt::Debug for App {
    fn fmt(&self, _f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Ok(())
    }
}

#[derive(Debug)]
pub struct Config {
    pub db_url: String,
}

#[derive(Clone)]
pub struct ServerState {
    pub app: &'static Arc<&'static App>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Book {
    pub id: i32,
    pub title: String,
}

#[derive(Deserialize, Debug)]
pub struct BookGet {
    pub title: String,
}

#[derive(Debug)]
pub struct BookGetMessage {
    pub channel: crossbeam_channel::Sender<Option<Book>>,
    pub payload: BookGet,
    pub app: &'static App,
}
