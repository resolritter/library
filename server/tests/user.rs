use entities::{access_level, user, UserCreationPayload};
use insta::assert_snapshot;
use stdext::function_name;
use surf::StatusCode;
use tempdir::TempDir;
use test_utils::{format::format_test_name, read_snapshot, spawn_test_program, SpawnedTest};

#[async_std::test]
async fn test_create_user() {
    let test_name = format_test_name(function_name!());
    let tmp_dir = TempDir::new(&test_name).unwrap();

    // Calling the test server with those credentials will create a superuser at the start
    let admin_access_token = "ADMIN";
    let admin_email = "admin@admin.com";
    let SpawnedTest {
        server_addr,
        log_dir,
        process: _,
    } = &spawn_test_program(
        &tmp_dir,
        Some(format!("{}::{}", admin_email, admin_access_token)),
    );

    user::create(
        &server_addr,
        &UserCreationPayload {
            email: "librarian@user.com".to_string(),
            access_level: access_level::LIBRARIAN,
            // Administrators can create librarians
            requester_access_token: Some(admin_access_token.to_string()),
        },
    )
    .await;
    let (_, simple_user) = user::create(
        &server_addr,
        &UserCreationPayload {
            email: "simple@user.com".to_string(),
            access_level: access_level::USER,
            // Normal user creation does not require any sort of special access
            requester_access_token: None,
        },
    )
    .await;

    // Normal users should not be able to create any sort of special user
    let bad_forbidden_request_admin = user::do_create(
        &server_addr,
        &UserCreationPayload {
            email: "new_ADMIN@user.com".to_string(),
            access_level: access_level::ADMIN,
            requester_access_token: Some(simple_user.access_token.clone()),
        },
    )
    .await;
    assert!(bad_forbidden_request_admin.status() == StatusCode::Forbidden);
    let bad_forbidden_request_librarian = user::do_create(
        &server_addr,
        &UserCreationPayload {
            email: "new_LIBRARIAN@user.com".to_string(),
            access_level: access_level::LIBRARIAN,
            requester_access_token: Some(simple_user.access_token.clone()),
        },
    )
    .await;
    assert!(bad_forbidden_request_librarian.status() == StatusCode::Forbidden);

    assert_snapshot!(read_snapshot(&log_dir));
}
