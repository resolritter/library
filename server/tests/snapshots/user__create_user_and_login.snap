---
source: server/tests/user.rs
expression: read_snapshot(&log_dir)
---
UserLoginPayload {
    email: "DOES_NOT_EXIST@user.com",
}
UserCreationPayload {
    email: "simple@user.com",
    access_mask: 1,
    requester_access_token: None,
}
UserLoginPayload {
    email: "simple@user.com",
}
